File: infer/tests/codetoanalyze/java/pulse/taint/Basics.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

/** testing basic intraprocedural functionality: assignment, ifs, loops, casts */
public class Basics {

  native Object notASource();

  native void notASink(Object o);

  /** should report on these tests */
  void directBad() {
    InferTaint.inferSensitiveSink(InferTaint.inferSecretSource());
  }

  void viaVarBad1() {
    Object src = InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(src);
  }

  void viaVarBad2() {
    Object src = InferTaint.inferSecretSource();
    Object alias = src;
    InferTaint.inferSensitiveSink(alias);
  }

  void viaVarBad3() {
    Object src = InferTaint.inferSecretSource();
    Object alias = src;
    src = null;
    InferTaint.inferSensitiveSink(alias);
  }

  void viaCastBad1() {
    InferTaint.inferSensitiveSink((String) InferTaint.inferSecretSource());
  }

  void viaCastBad2() {
    Object src = InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink((String) src);
  }

  void ifBad1(boolean b) {
    Object src = null;
    if (b) {
      src = InferTaint.inferSecretSource();
    }
    InferTaint.inferSensitiveSink(src);
  }

  void ifBad2(boolean b) {
    Object src = InferTaint.inferSecretSource();
    if (b) {
      src = null;
    }
    InferTaint.inferSensitiveSink(src);
  }

  void ifBad3(boolean b) {
    Object src;
    if (b) {
      src = new Object();
    } else {
      src = InferTaint.inferSecretSource();
    }
    InferTaint.inferSensitiveSink(src);
  }

  void ifBad4(boolean b1, boolean b2) {
    Object src;
    if (b1) {
      src = new Object();
    } else if (b2) {
      src = InferTaint.inferSecretSource();
    } else {
      src = null;
    }
    InferTaint.inferSensitiveSink(src);
  }

  void ifBad5(boolean b) {
    Object src = InferTaint.inferSecretSource();
    if (b) {
      InferTaint.inferSensitiveSink(src);
    }
  }

  void switchBad1(int i) {
    Object src = InferTaint.inferSecretSource();
    switch (i) {
      case 1:
        InferTaint.inferSensitiveSink(src);
        break;
      case 2:
        break;
      default:
        break;
    }
  }

  void switchBad2(int i) {
    Object src = InferTaint.inferSecretSource();
    switch (i) {
      case 1:
        break;
      case 2:
        InferTaint.inferSensitiveSink(src);
        break;
      default:
        break;
    }
  }

  void switchBad3(int i) {
    Object src = null;
    switch (i) {
      case 1:
        src = InferTaint.inferSecretSource();
        // fallthrough
      case 2:
        InferTaint.inferSensitiveSink(src);
        break;
      default:
        break;
    }
  }

  void whileBad1(int i) {
    Object src = InferTaint.inferSecretSource();
    while (i < 10) {
      InferTaint.inferSensitiveSink(src);
      i++;
    }
  }

  void whileBad2(int i) {
    Object src = null;
    while (i < 10) {
      src = InferTaint.inferSecretSource();
      i++;
    }
    InferTaint.inferSensitiveSink(src);
  }

  // this should report only two alarms, not three
  void noTripleReportBad() {
    Object src = InferTaint.inferSecretSource();
    InferTaint.inferSensitiveSink(src);
    InferTaint.inferSensitiveSink(src);
  }

  void arrayWithTaintedContentsBad() {
    Object src = InferTaint.inferSecretSource();
    Object[] arr = new Object[] {src};
    InferTaint.inferSensitiveSink(arr);
  }

  void passToSinkOk(int x, Object src) {
    InferTaint.inferSensitiveSink(src);
  }

  void funCallBad() {
    Object src = InferTaint.inferSecretSource();
    passToSinkOk(2, src);
  }

  /** should not report on these tests */
  void directOk1() {
    notASink(notASource());
  }

  void directOk2() {
    notASink(InferTaint.inferSecretSource());
  }

  void directOk3() {
    InferTaint.inferSensitiveSink(notASource());
  }

  void viaVarOk() {
    Object src = new Object();
    InferTaint.inferSensitiveSink(src);
  }

  void viaVarStrongUpdateOk() {
    Object src = InferTaint.inferSecretSource();
    src = null;
    InferTaint.inferSensitiveSink(src);
  }

  Object exceptionOk(boolean b, Object o) {
    if (b) {
      throw new AssertionError("exception");
    }
    o.toString();
    return o;
  }

  void synchronizedOk(Object o) {
    synchronized (o) {
    }
  }

  // this is to test that we don't crash due to the slightly odd translation of synchronized
  void callSynchronizedOk(Object o) {
    synchronizedOk(o);
  }

  void deadCodeOk() {
    Object src = InferTaint.inferSecretSource();
    boolean b = false;
    if (b) {
      InferTaint.inferSensitiveSink(src);
    }
  }

  void loopInvariantOk() {
    Object src = InferTaint.inferSecretSource();
    for (int i = 0; i < 10; i++) {
      src = null;
    }
    InferTaint.inferSensitiveSink(src);
  }

  void taintedToSanitizedToSinkOk() {
    InferTaint.inferSensitiveSink(
        InferTaint.inferUniversalSanitizer(InferTaint.inferSecretSource()));
  }

  Object inferUniversalSanitizer() {
    return InferTaint.inferSecretSource();
  }

  // sanitizer should take precedence
  void FP_returnedFromSanitizedTaintedToSinkOk() {
    InferTaint.inferSensitiveSink(inferUniversalSanitizer());
  }

  void taintOnUnrelatedBoolOk(boolean notTaintedFlag) {
    boolean taintedFlag = (boolean) InferTaint.inferSecretSource();
    boolean uberFlag = notTaintedFlag || taintedFlag;
    if (!uberFlag) {
      InferTaint.inferSensitiveSink(notTaintedFlag);
    }
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/Basics.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

/** testing basic intraprocedural functionality: assignment, ifs, loops, casts */
class Basics {

  external fun notASource(): Any

  external fun notASink(o: Any)

  /** should report on these tests */
  fun directBad() {
    InferTaint.inferSensitiveSink(InferTaint.inferSecretSource())
  }

  fun viaVarBad1() {
    val src = InferTaint.inferSecretSource()
    InferTaint.inferSensitiveSink(src)
  }

  fun viaVarBad2() {
    val src = InferTaint.inferSecretSource()
    InferTaint.inferSensitiveSink(src)
  }

  fun viaVarBad3() {
    var src: Any? = InferTaint.inferSecretSource()
    val alias = src as Any
    src = null
    InferTaint.inferSensitiveSink(alias)
  }

  fun viaCastBad1() {
    InferTaint.inferSensitiveSink((InferTaint.inferSecretSource() as String))
  }

  fun viaCastBad2() {
    val src = InferTaint.inferSecretSource()
    InferTaint.inferSensitiveSink((src as String))
  }

  fun ifBad1(b: Boolean) {
    var src: Any? = null
    if (b) {
      src = InferTaint.inferSecretSource()
    }
    InferTaint.inferSensitiveSink(src)
  }

  fun ifBad2(b: Boolean) {
    var src: Any? = InferTaint.inferSecretSource()
    if (b) {
      src = null
    }
    InferTaint.inferSensitiveSink(src)
  }

  fun ifBad3(b: Boolean) {
    val src: Any =
        if (b) {
          Any()
        } else {
          InferTaint.inferSecretSource()
        }
    InferTaint.inferSensitiveSink(src)
  }

  fun ifBad4(b1: Boolean, b2: Boolean) {
    val src: Any? =
        if (b1) {
          Any()
        } else if (b2) {
          InferTaint.inferSecretSource()
        } else {
          null
        }
    InferTaint.inferSensitiveSink(src)
  }

  fun ifBad5(b: Boolean) {
    val src = InferTaint.inferSecretSource()
    if (b) {
      InferTaint.inferSensitiveSink(src)
    }
  }

  fun switchBad1(i: Int) {
    val src = InferTaint.inferSecretSource()
    when (i) {
      1 -> InferTaint.inferSensitiveSink(src)
      2 -> {}
      else -> {}
    }
  }

  fun switchBad2(i: Int) {
    val src = InferTaint.inferSecretSource()
    when (i) {
      1 -> {}
      2 -> InferTaint.inferSensitiveSink(src)
      else -> {}
    }
  }

  fun whileBad1(input: Int) {
    var i = input
    val src = InferTaint.inferSecretSource()
    while (i < 10) {
      InferTaint.inferSensitiveSink(src)
      i++
    }
  }

  fun whileBad2(input: Int) {
    var i = input
    var src: Any? = null
    while (i < 10) {
      src = InferTaint.inferSecretSource()
      i++
    }
    InferTaint.inferSensitiveSink(src)
  }

  // this should report only two alarms, not three
  fun noTripleReportBad() {
    val src = InferTaint.inferSecretSource()
    InferTaint.inferSensitiveSink(src)
    InferTaint.inferSensitiveSink(src)
  }

  fun arrayWithTaintedContentsBad() {
    val src = InferTaint.inferSecretSource()
    val arr = arrayOf(src)
    InferTaint.inferSensitiveSink(arr)
  }

  fun passToSinkOk(x: Int, src: Any?) {
    InferTaint.inferSensitiveSink(src)
  }

  fun funCallBad() {
    val src = InferTaint.inferSecretSource()
    passToSinkOk(2, src)
  }

  /** should not report on these tests */
  fun directOk1() {
    notASink(notASource())
  }

  fun directOk2() {
    notASink(InferTaint.inferSecretSource())
  }

  fun directOk3() {
    InferTaint.inferSensitiveSink(notASource())
  }

  fun viaVarOk() {
    val src = Any()
    InferTaint.inferSensitiveSink(src)
  }

  fun viaVarStrongUpdateOk() {
    var src: Any? = InferTaint.inferSecretSource()
    src = null
    InferTaint.inferSensitiveSink(src)
  }

  fun exceptionOk(b: Boolean, o: Any): Any {
    if (b) {
      throw AssertionError("exception")
    }
    o.toString()
    return o
  }

  fun synchronizedOk(o: Any) {
    synchronized(o) {}
  }

  // this is to test that we don't crash due to the slightly odd translation of synchronized
  fun callSynchronizedOk(o: Any) {
    synchronizedOk(o)
  }

  fun deadCodeOk() {
    val src = InferTaint.inferSecretSource()
    val b = false
    if (b) {
      InferTaint.inferSensitiveSink(src)
    }
  }

  fun loopInvariantOk() {
    var src: Any? = InferTaint.inferSecretSource()
    for (i in 0..9) {
      src = null
    }
    InferTaint.inferSensitiveSink(src)
  }

  fun taintedToSanitizedToSinkOk() {
    InferTaint.inferSensitiveSink(
        InferTaint.inferUniversalSanitizer(InferTaint.inferSecretSource()))
  }

  fun taintOnUnrelatedBooleanOk(notTaintedFlag: Boolean) {
    val taintedFlag = InferTaint.inferSecretSource() as Boolean
    val uberFlag = notTaintedFlag || taintedFlag
    if (!uberFlag) {
      InferTaint.inferSensitiveSink(notTaintedFlag)
    }
  }

  fun javaSourceBad() {
    val src = InferJavaTaint.inferSecretSource()
    InferTaint.inferSensitiveSink(src)
  }

  fun javaSinkBad() {
    val src = InferTaint.inferSecretSource()
    InferJavaTaint.inferSensitiveSink(src)
  }
}
